### DISCUSSION TITLE
Introduce optional(?) buildscripts that generate .deb and .rpm files

### DISCUSSION THREAD

**Discussion by ghost** [OpenBangla/OpenBangla-Keyboard#367] [Ideas]
Currently the suggested method for installation for everyone seems to be with the [install.sh](https://github.com/OpenBangla/OpenBangla-Keyboard/blob/master/tools/install.sh) file, which downloads the appropriate .deb/.rpm/.pkg file and installs it accordingly, or by manually compiling and installing it by `make install`

Distros like Ubuntu XX.10, Debian Testing and Fedora continually update and newer versions of dependencies get introduced, which might conflict with the pre-compiled binaries from github actions.
This situation warrants that the user manually compile OBK. The user might also want to use OBK from the develop branch, which also requires manual compilation.

`make install` installs OBK manually to /usr/local/ and it stays unmanaged by the package manager. I find it is better in linux if all software is installed via the package manager, so everything stays centralized, the user can easily and cleanly un/reinstall OBK and additional checks may be performed by the package manager to ensure any missing/outdated dependencies are installed/updated.

My proposal is to include two optional `make-deb.sh` and `make-rpm.sh` files, which compile OBK from source and generate .deb and .rpm files respectively, and additionally allow the user to specify which branch/version of OBK they want.
The scripts might potentially check for any missing dependencies, install them accordingly and optionally run the compilation task inside a toolbox/distrobox container so the host system stays untouched.

Benefits would be that instead of having to run several commands the user might potentially be unfamiliar with, or avoiding the risk of messing up his system with build dependencies, the user will run a simple single .sh file (potentially with some args) and have a .deb/.rpm file he can double-click to easily install via a GUI package manager.

I am willing to work on this, but I would like to hear the opinion of others.

**Comment by mominul** [comment]
Thanks, I appreciate the thought!

I am currently working on a tool that will enable users to install OBK right through their package manager. But if that fails we'll need to follow a path similar to what you have described.

I have initially tested that from Debian 11 and Ubuntu 20.04, the dependencies are the same in their new versions. I have only tested if the dependencies satisfy using docker, but thorough testing is needed.

Can you test Debian (from 11) and Ubuntu (from 20.04) distributions and their derivatives (ZorinOS, Deepin, Linux Mint) using [these two `deb` files](https://drive.google.com/drive/folders/1Ta6-pPTbqHdKLfN15LmDD46sX3kNzIyK?usp=sharing)?

About the IM engines, test ibus engines on Gnome and other Gnome derivatives (I think those still prefer ibus, though I am not sure about Mint or Cinnamon), fcitx5 on KDE, and Deepin. Will have to check what provides the best integration on ZorinOS too.

That would help me a ton! Thanks in advance! :heart:

**Reply by ghost** [reply]
> I am currently working on a tool that will enable users to install OBK right through their package manager.

Honestly, the distribution issue could easily be solved if OBK was merged into the Debian and Fedora repos, it would trickle down to all other downstream forks like Ubuntu/LM/Zorin and Nobara/uBlue etc.
On the otherhand, most distros ship user-friendly package installers for .deb and .rpm files, Debian has gdebi for example.
IMO handing people .deb/.rpm files would be more than enough, as most people are accustomed to installing software via .msi and .exe installers in Windows.

**Reply by ghost** [reply]
I am on a 240GB SSD laptop, Sadly I can only test Fedora (which I use), Debian 12/Ubuntu 22.04 LTS and Linux Mint 21.X (inside VMs) due to lack of space

**Reply by mominul** [reply]
> I am on a 240GB SSD laptop, Sadly I can only test Fedora (which I use), Debian 12/Ubuntu 22.04 LTS and Linux Mint 21.X (inside VMs) due to lack of space

I intended you to use VMs to test these distributions. :3 You only have to check if OBK runs properly (and the TopBar opens) and document the DE integrations for #368

**Reply by ghost** [reply]
Sadly, the testing will have to be postponed for sometime on my end. My laptop's SSD died yesterday night, I'll need to replace it before I can do anything else.

**Reply by mominul** [reply]
I have also faced SSD issues on my PC the last couple of months, so I know how frustrating it can be. :cry:

**Reply by ghost** [reply]
Replaced my SSD, going to spin up the VMs
Here's my list
- Debian 11
- Debian 12
- Ubuntu 20.04 LTS
- Ubuntu 22.04 LTS
- Linux Mint 21.2

Although I had to ask, Ubuntu 22.04 LTS has been out for almost 2 years, Debian 12 for 6+ months, seems redundant testing on older releases.

**Reply by ghost** [reply]
@mominul I have tested OBK on

1. Linux Mint 21
- has neither ibus nor fcitx installed/setup ootb
- can use either
- needs ibus-setup and fcitx5-configtool to manually enable OBK, the GUI menu appears to be broken
- both ibus and fcitx OBK versions work fine, the obk .deb files can be installed via `sudo apt install <OBK>.deb`

2. Debian 11
- has ibus installed and setup ootb on GNOME
- ibus-obk installs fine, works fine, but fonts are "empty block characters", indicating missing bangla fonts
- have tried doing `sudo apt install fonts-noto\*` which should hopefully fix the missing fonts issue, but due to ISPs throttling internet bandwidth today nationwide, couldn't check for sure, will check at a future date
- will check fcitx-obk on KDE in Debian 11/12 at a future date

3. Ubuntu 20.04
- has ibus installed and setup ootb on GNOME
- ibus-obk installs with debian instructions, works fine after install, no font issues

4. Ubuntu 22.04
- has ibus installed and setup ootb on GNOME
- ibus-obk install fails with cryptic error message, same as https://github.com/OpenBangla/OpenBangla-Keyboard/issues/369

![Screenshot from 2024-01-07 22-59-07](https://github.com/OpenBangla/OpenBangla-Keyboard/assets/138502262/cadfce43-716d-4672-a963-4a9f904dbcd9)

**Reply by ghost** [reply]
let me know if you want KDE to be tested on Ubuntu too

**Reply by ghost** [reply]
As the internet situation has improved, I have tested Debian and Ubuntu too,

- Debian 11
    - ibus works fine
    - sudo apt install fonts-noto\* does fix the broken fonts
    - fcitx is broken, tested in KDE on Firefox and Kwrite to rule out any GTK/Qt issue
    
    On ibus/gnome :
    
![Screenshot from 2024-01-08 00-49-54](https://github.com/OpenBangla/OpenBangla-Keyboard/assets/138502262/67f143bc-51ce-43c4-8fbf-7400c82b5ae1)

    on fcitx/kde :

[Screencast from 2024-01-08 01-03-18.webm](https://github.com/OpenBangla/OpenBangla-Keyboard/assets/138502262/a56bc4ca-4fe6-4511-9f79-d75e150fa0cf)


- Debian 12
    - the fonts-noto packages are already preinstalled, no need to manually install
    - attempt to install ibus-obk errors out with the same error as ubuntu
    
    
![Screenshot from 2024-01-08 01-12-34](https://github.com/OpenBangla/OpenBangla-Keyboard/assets/138502262/c2971c70-1aa5-4852-87b4-13535de7ca75)

**Reply by ghost** [reply]
@mominul closing the discussion as per https://github.com/OpenBangla/OpenBangla-Keyboard/discussions/367#discussioncomment-8033023
Further testing warrants an Issue in the Issues section.

**Reply by mominul** [reply]
@larina3315 Thanks a lot! About the cryptic error message, I think the installation might have succeeded, have you checked if the binaries of OBK are present?

**Reply by ghost** [reply]
> have you checked if the binaries of OBK are present?

You mean under /usr ? no I did not, I assumed apt reverted any changes it had made after the error message

**Reply by mominul** [reply]
I meant it because I think that message is a warning, not an error.

**Reply by ghost** [reply]
I will need to recheck again when I can afford the time, but IIRC I didn't see any desktop (app menu) entry for OBK after that message, so i assumed the install failed.

**Comment by ghost** [comment]
I have created the necessary shell files for generating .rpms on fedora
support can easily be extended for other distros by manually editing the file
https://github.com/OpenBangla/OpenBangla-Keyboard/pull/370

I will document how I intend for it to work.

1. User can specify --ibus or --fcitx to generate ibus and fcitx versions of obk respectively (both flags can be used together to generate both ibus and fcitx versions of OBK together for a given distro.)
2. User can specify distro name, like --fedora or --debian to generate for that specific distro (both flags can be used together to generate binaries for specified distros together, ONLY if the --toolbox flag is set)
They can also specify version e.g. --fedora 38 or --debian 11
3. User can specify --toolbox flag to compile/clone inside of a toolbox (not mentioning the flag runs the compile/clone etc tasks on users host system)
4. User can specify --develop to compile from the develop branch of OBK (Otherwise uses the master branch)

example usage : ./make-pkgs.sh --fedora 38 --ibus --fcitx --develop --toolbox
compiles the ibus version of OBK from the develop branch for both debian (.deb) and fedora (.rpm) using toolbox containers

Currently debian support is not implemented, but can easily be done so by adding a `make-debian.sh` file modified from the `make-fedora.sh' file
It still has some issues, so it is not fit for end-user usage.

**Reply by ghost** [reply]
TODO : 
- (DONE) Add check to make sure user doesn't try to compile for multiple distros without the --toolbox flag
- Disable fcitx flag on OBK2 and earlier
- (DONE) Allow user to specify the version of the distro they want to compile in (--toolbox only)
- (DONE) Add --help file and proper docs
- Allow user to select a branch other than develop

**Comment by ghost** [comment]
The script is now finished, I will close the issue after testing https://github.com/OpenBangla/OpenBangla-Keyboard/discussions/367#discussioncomment-7988670
